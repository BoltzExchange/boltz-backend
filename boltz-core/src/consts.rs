pub const ECDSA_BYTES_TO_GRIND: usize = 2;

pub const PREIMAGE_DUMMY: [u8; 0] = [];

pub const STUB_SCHNORR_SIGNATURE_LENGTH: usize = 64;
