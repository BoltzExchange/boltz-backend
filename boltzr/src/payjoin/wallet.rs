
use anyhow::Result;
use payjoin::bitcoin::psbt::Psbt;
use payjoin::bitcoin::{
    Address, Network, Script, Transaction, Txid,
};
use payjoin::receive::InputPair;

/// Implementation of PayjoinWallet for bitcoind
#[derive(Clone, Debug)]
pub struct BitcoindWallet {

}

impl BitcoindWallet {
    pub fn new() -> Result<Self> {
        Ok(Self {})
    }
}

impl BitcoindWallet {
    /// Process a PSBT, validating and signing inputs owned by this wallet
    ///
    /// Does not include bip32 derivations in the PSBT
    pub fn process_psbt(&self, psbt: &Psbt) -> Result<Psbt> {
        let psbt_str = psbt.to_string();
        todo!("wallet process psbt");
    }

    /// Finalize a PSBT and extract the transaction
    pub fn finalize_psbt(&self, psbt: &Psbt) -> Result<Transaction> {
        todo!("wallet finalize psbt");
    }

    /// Check if a transaction can be broadcast
    pub fn can_broadcast(&self, tx: &Transaction) -> Result<bool> {
        // FIXME
        Ok(true)
    }

    /// Broadcast a raw transaction
    pub fn broadcast_tx(&self, tx: &Transaction) -> Result<Txid> {
        todo!("wallet broadcast tx");
    }

    /// Check if a script belongs to this wallet
    pub fn is_mine(&self, script: &Script) -> Result<bool> {
        if let Ok(_address) = Address::from_script(script, self.network()?) {
           // FIXME
           Ok(true)
        } else {
            Ok(false)
        }
    }

    /// Get a new address from the wallet
    pub fn get_new_address(&self) -> Result<Address> {
        todo!("get new address")
    }

    /// List unspent UTXOs
    pub fn list_unspent(&self) -> Result<Vec<InputPair>> {
        // FIXME
        Ok(vec![])
    }

    /// Get the network this wallet is operating on
    pub fn network(&self) -> Result<Network> {
        todo!("get network")
    }
}

pub fn input_pair_from_list_unspent(
    // utxo: bitcoincore_rpc::bitcoincore_rpc_json::ListUnspentResultEntry,
) -> InputPair {
    unimplemented!("input pair from list unspent")
}
