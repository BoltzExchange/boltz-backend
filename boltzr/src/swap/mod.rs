mod expiry_checker;
mod filters;
pub mod manager;
mod status;

pub use status::*;
