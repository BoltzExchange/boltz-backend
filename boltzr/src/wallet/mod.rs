use ::bitcoin::bip32::Xpriv;
use anyhow::Result;
use async_trait::async_trait;

mod bitcoin;
mod elements;
mod keys;

pub use bitcoin::*;
pub use elements::*;

#[derive(PartialEq, Debug, Clone, Copy)]
pub enum Network {
    Mainnet,
    Testnet,
    Signet,
    Regtest,
}

#[async_trait]
pub trait Wallet {
    fn decode_address(&self, address: &str) -> Result<Vec<u8>>;
    fn derive_keys(&self, index: u64) -> Result<Xpriv>;
    fn derive_blinding_key(&self, address: &str) -> Result<Vec<u8>>;

    async fn get_address(&self, label: String) -> Result<String>;
}

impl Network {
    pub fn bitcoin(&self) -> ::bitcoin::Network {
        match self {
            Network::Mainnet => ::bitcoin::Network::Bitcoin,
            Network::Signet => ::bitcoin::Network::Signet,
            Network::Testnet => ::bitcoin::Network::Testnet,
            Network::Regtest => ::bitcoin::Network::Regtest,
        }
    }

    pub fn liquid(&self) -> anyhow::Result<&'static ::elements::AddressParams> {
        match self {
            Network::Mainnet => Ok(&::elements::address::AddressParams::LIQUID),
            Network::Testnet => Ok(&::elements::address::AddressParams::LIQUID_TESTNET),
            Network::Regtest => Ok(&::elements::address::AddressParams::ELEMENTS),
            Network::Signet => Err(anyhow::anyhow!(
                "Signet is not supported for liquid addresses"
            )),
        }
    }
}

#[cfg(test)]
pub mod test {
    use super::*;

    pub use keys::test::get_seed;
}
