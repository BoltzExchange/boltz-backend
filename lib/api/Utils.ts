import { Request, Response } from 'express';
import Logger from '../Logger';
import { getHexBuffer } from '../Utils';
import MarkedSwapRepository from '../db/repositories/MarkedSwapRepository';
import CountryCodes from '../service/CountryCodes';
import ServiceErrors from '../service/Errors';
import Errors from './Errors';

type ApiArgument = {
  name: string;
  type: 'string' | 'number' | 'boolean' | 'object';
  hex?: boolean;
  optional?: boolean;
};

// Some endpoints are getting spammed on mainnet, so we don't log the warnings for them
const errorsNotToLog: any[] = [
  ServiceErrors.SWAP_NO_LOCKUP().message,
  ServiceErrors.ETHEREUM_NOT_ENABLED().message,
];

/**
 * Validates that all required arguments were provided in the body correctly
 *
 * @returns the validated arguments
 */
export const validateRequest = (
  body: Record<string, any>,
  argsToCheck: ApiArgument[],
) => {
  const response: any = {};

  argsToCheck.forEach((arg) => {
    const value = body[arg.name];

    if (value !== undefined) {
      if (typeof value === arg.type) {
        if (arg.hex && value !== '') {
          const buffer = getHexBuffer(value);

          if (buffer.length === 0) {
            throw Errors.COULD_NOT_PARSE_HEX(arg.name);
          }

          response[arg.name] = buffer;
        } else {
          response[arg.name] = value;
        }
      } else {
        throw Errors.INVALID_PARAMETER(arg.name);
      }
    } else if (!arg.optional) {
      throw Errors.UNDEFINED_PARAMETER(arg.name);
    }
  });

  return response;
};

export const errorResponse = (
  logger: Logger,
  req: Request,
  res: Response,
  error: unknown,
  statusCode = 400,
): void => {
  if (typeof error === 'string') {
    writeErrorResponse(logger, req, res, statusCode, { error });
  } else {
    const errorObject = error as any;

    // Bitcoin Core related errors
    if (errorObject.details) {
      writeErrorResponse(logger, req, res, statusCode, {
        error: errorObject.details,
      });
      // Custom error when broadcasting a refund transaction fails because
      // the locktime requirement has not been met yet
    } else if (errorObject.timeoutBlockHeight) {
      writeErrorResponse(logger, req, res, statusCode, error);
      // Everything else
    } else {
      writeErrorResponse(logger, req, res, statusCode, {
        error: errorObject.message,
      });
    }
  }
};

export const successResponse = (res: Response, data: unknown) => {
  setContentTypeJson(res);
  res.status(200);

  if (typeof data === 'object') {
    res.json(data);
  } else {
    res.write(data);
    res.end();
  }
};

export const createdResponse = (res: Response, data: unknown) => {
  setContentTypeJson(res);
  res.status(201).json(data);
};

export const writeErrorResponse = (
  logger: Logger,
  req: Request,
  res: Response,
  statusCode: number,
  error: any,
) => {
  if (!errorsNotToLog.includes(error?.error || error)) {
    logger.warn(
      `Request ${req.method} ${req.originalUrl} ${
        req.body && Object.keys(req.body).length > 0
          ? `${JSON.stringify(req.body)} `
          : ''
      }failed: ${JSON.stringify(error)}`,
    );
  }

  setContentTypeJson(res);
  res.status(statusCode).json(error);
};

export const setContentTypeJson = (res: Response) => {
  res.set('Content-Type', 'application/json');
};

export const checkPreimageHashLength = (preimageHash: Buffer) => {
  if (preimageHash.length !== 32) {
    throw `invalid preimage hash length: ${preimageHash.length}`;
  }
};

export const markSwap = async (
  countryCodes: CountryCodes,
  ip: string,
  swapId: string,
) => {
  if (!countryCodes.isRelevantCountry(countryCodes.getCountryCode(ip))) {
    return;
  }

  await MarkedSwapRepository.addMarkedSwap(swapId);
};
