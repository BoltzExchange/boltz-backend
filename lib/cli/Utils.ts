import { Arguments } from 'yargs';
import { address, ECPair, Transaction } from 'bitcoinjs-lib';
import { Networks, detectSwap, constructClaimTransaction, constructRefundTransaction } from 'boltz-core';
import inquire from './inquire';
import { getHexBuffer } from '../Utils';
import { OutputType, OrderSide } from '../proto/boltzrpc_pb';

export const getOrderSide = (side: string) => {
  switch (side.toLowerCase()) {
    case 'buy': return OrderSide.BUY;
    case 'sell': return OrderSide.SELL;

    default: throw `could not find order side: ${side}`;
  }
};

export const getOutputType = (type: string) => {
  switch (type.toLowerCase()) {
    case 'bech32': return OutputType.BECH32;
    case 'compatibility': return OutputType.COMPATIBILITY;

    default: return OutputType.LEGACY;
  }
};

export const getNetwork = (networkKey: string) => {
  const network = Networks[networkKey];

  if (!network) {
    throw `Could not find network: ${networkKey}`;
  }

  return network;
};

const parseParameters = (argv: Arguments) => {
  const network = getNetwork(argv.network);

  return {
    network,
    lockupTransaction: Transaction.fromHex(argv.lockup_transaction),
    redeemScript: getHexBuffer(argv.redeem_script),
    destinationScript: address.toOutputScript(argv.destination_address, network),
  };
};

const parseSwapOutput = (redeemScript: Buffer, lockupTransaction: Transaction) => {
  const swapOutput = detectSwap(redeemScript, lockupTransaction);

  if (!swapOutput) {
    throw 'Could not find swap in transaction';
  }

  return swapOutput;
};

export const parseCommands = async (inquiries: any[], argv: Arguments): Promise<Arguments> => {
  const argvLength = Object.keys(argv).length;
  if (argvLength === inquiries.length) {
    const answers = await inquire(inquiries);
    return { ...answers, ...argv };
  } else {
    return argv;
  }
};

export const claimSwap = (argv: Arguments) => {
  const { network, lockupTransaction, redeemScript, destinationScript } = parseParameters(argv);
  const claimKeys = ECPair.fromPrivateKey(getHexBuffer(argv.claim_private_key), { network });

  const swapOutput = parseSwapOutput(redeemScript, lockupTransaction);

  const claimTransaction = constructClaimTransaction(
    {
      redeemScript,
      preimage: getHexBuffer(argv.preimage),
      keys: claimKeys,
    },
    {
      txHash: lockupTransaction.getHash(),
      ...swapOutput,
    },
    destinationScript,
    argv.fee_per_byte,
  );

  return claimTransaction.toHex();
};

export const refundSwap = (argv: Arguments) => {
  const { network, lockupTransaction, redeemScript, destinationScript } = parseParameters(argv);
  const refundKeys = ECPair.fromPrivateKey(getHexBuffer(argv.refund_private_key), { network });

  const swapOutput = parseSwapOutput(redeemScript, lockupTransaction);

  const refundTransaction = constructRefundTransaction(
    refundKeys,
    redeemScript,
    argv.timeout_block_height,
    {
      txHash: lockupTransaction.getHash(),
      ...swapOutput,
    },
    destinationScript,
    argv.fee_per_byte,
  );

  return refundTransaction.toHex();
};
