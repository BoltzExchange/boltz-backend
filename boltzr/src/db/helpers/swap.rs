use crate::db::helpers::{BoxedCondition, QueryResponse};
use crate::db::models::Swap;
use crate::db::schema::swaps;
use crate::db::Pool;
use crate::swap::SwapUpdate;
use diesel::prelude::*;
use diesel::{update, QueryDsl, RunQueryDsl, SelectableHelper};

pub type SwapCondition = BoxedCondition<swaps::table>;

pub trait SwapHelper {
    fn get_all(&self, condition: SwapCondition) -> QueryResponse<Vec<Swap>>;
    fn update_status(
        &self,
        id: &str,
        status: SwapUpdate,
        failure_reason: Option<String>,
    ) -> QueryResponse<usize>;
}

#[derive(Clone, Debug)]
pub struct SwapHelperDatabase {
    pool: Pool,
}

impl SwapHelperDatabase {
    pub fn new(pool: Pool) -> Self {
        Self { pool }
    }
}

impl SwapHelper for SwapHelperDatabase {
    fn get_all(&self, condition: BoxedCondition<swaps::dsl::swaps>) -> QueryResponse<Vec<Swap>> {
        Ok(swaps::dsl::swaps
            .select(Swap::as_select())
            .filter(condition)
            .load(&mut self.pool.get()?)?)
    }

    fn update_status(
        &self,
        id: &str,
        status: SwapUpdate,
        failure_reason: Option<String>,
    ) -> QueryResponse<usize> {
        if let Some(failure_reason) = failure_reason {
            Ok(update(swaps::dsl::swaps)
                .filter(swaps::dsl::id.eq(id))
                .set((
                    swaps::dsl::status.eq(status.to_string()),
                    swaps::dsl::failureReason.eq(failure_reason),
                ))
                .execute(&mut self.pool.get()?)?)
        } else {
            Ok(update(swaps::dsl::swaps)
                .filter(swaps::dsl::id.eq(id))
                .set(swaps::dsl::status.eq(status.to_string()))
                .execute(&mut self.pool.get()?)?)
        }
    }
}
