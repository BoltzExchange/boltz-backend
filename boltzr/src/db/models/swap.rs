use crate::db::models::{LightningSwap, SomeSwap, SwapType};
use crate::swap::SwapUpdate;
use crate::utils::pair::{split_pair, OrderSide};
use diesel::{AsChangeset, Insertable, Queryable, Selectable};

#[derive(Queryable, Selectable, Insertable, AsChangeset, PartialEq, Default, Clone, Debug)]
#[diesel(table_name = crate::db::schema::swaps)]
#[allow(non_snake_case)]
pub struct Swap {
    pub id: String,
    pub referral: Option<String>,
    pub pair: String,
    pub orderSide: i32,
    pub status: String,
    pub failureReason: Option<String>,
    pub invoice: Option<String>,
    pub lockupAddress: String,
    pub createdAt: chrono::NaiveDateTime,
}

impl SomeSwap for Swap {
    fn kind(&self) -> SwapType {
        SwapType::Submarine
    }

    fn id(&self) -> String {
        self.id.clone()
    }

    fn status(&self) -> SwapUpdate {
        SwapUpdate::parse(self.status.as_str())
    }
}

impl LightningSwap for Swap {
    fn chain_symbol(&self) -> anyhow::Result<String> {
        let pair = split_pair(&self.pair)?;
        Ok(if self.orderSide == OrderSide::Buy as i32 {
            pair.quote
        } else {
            pair.base
        })
    }

    fn lightning_symbol(&self) -> anyhow::Result<String> {
        let pair = split_pair(&self.pair)?;
        Ok(if self.orderSide == OrderSide::Buy as i32 {
            pair.base
        } else {
            pair.quote
        })
    }
}

#[cfg(test)]
mod test {
    use crate::db::models::{LightningSwap, SomeSwap, Swap, SwapType};
    use crate::swap::SwapUpdate;
    use crate::utils::pair::OrderSide;
    use rstest::*;

    #[test]
    fn test_kind() {
        assert_eq!(create_swap(None).kind(), SwapType::Submarine);
    }

    #[test]
    fn test_id() {
        let swap = create_swap(None);
        assert_eq!(swap.id(), swap.id);
    }

    #[test]
    fn test_status() {
        let swap = create_swap(None);
        assert_eq!(swap.status(), SwapUpdate::parse(swap.status.as_str()))
    }

    #[rstest]
    #[case(OrderSide::Buy, "BTC")]
    #[case(OrderSide::Sell, "L-BTC")]
    fn test_chain_symbol(#[case] side: OrderSide, #[case] expected: &str) {
        let swap = create_swap(Some(side));
        assert_eq!(swap.chain_symbol().unwrap(), expected);
    }

    #[rstest]
    #[case(OrderSide::Buy, "L-BTC")]
    #[case(OrderSide::Sell, "BTC")]
    fn test_lightning_symbol(#[case] side: OrderSide, #[case] expected: &str) {
        let swap = create_swap(Some(side));
        assert_eq!(swap.lightning_symbol().unwrap(), expected);
    }

    fn create_swap(order_side: Option<OrderSide>) -> Swap {
        Swap {
            id: "swap id".to_string(),
            pair: "L-BTC/BTC".to_string(),
            status: "transaction.mempool".to_string(),
            orderSide: order_side.unwrap_or(OrderSide::Buy) as i32,
            ..Default::default()
        }
    }
}
