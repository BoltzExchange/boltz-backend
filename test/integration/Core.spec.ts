import { BIP32Factory } from 'bip32';
import { generateMnemonic, mnemonicToSeedSync } from 'bip39';
import { Transaction, address, crypto } from 'bitcoinjs-lib';
import { toXOnly } from 'bitcoinjs-lib/src/psbt/bip371';
import {
  ClaimDetails,
  Networks,
  OutputType,
  Scripts,
  SwapTreeSerializer,
  detectSwap,
  swapScript,
  swapTree,
} from 'boltz-core';
import { Networks as LiquidNetworks } from 'boltz-core/dist/lib/liquid';
import { p2trOutput, p2wshOutput } from 'boltz-core/dist/lib/swap/Scripts';
import { randomBytes } from 'crypto';
import {
  Creator,
  CreatorInput,
  Extractor,
  Finalizer,
  Transaction as LiquidTransaction,
  Updater,
  address as liquidAddress,
  networks,
} from 'liquidjs-lib';
import { SLIP77Factory } from 'slip77';
import * as ecc from 'tiny-secp256k1';
import {
  constructClaimDetails,
  createMusig,
  fromOutputScript,
  getOutputValue,
  hashForWitnessV1,
  parseTransaction,
  setup,
  toOutputScript,
  tweakMusig,
} from '../../lib/Core';
import { ECPair } from '../../lib/ECPairHelper';
import Logger from '../../lib/Logger';
import { getHexBuffer, getHexString } from '../../lib/Utils';
import { CurrencyType, SwapVersion } from '../../lib/consts/Enums';
import Database from '../../lib/db/Database';
import Swap from '../../lib/db/models/Swap';
import SwapOutputType from '../../lib/swap/SwapOutputType';
import Wallet from '../../lib/wallet/Wallet';
import WalletLiquid from '../../lib/wallet/WalletLiquid';
import { Currency } from '../../lib/wallet/WalletManager';
import CoreWalletProvider from '../../lib/wallet/providers/CoreWalletProvider';
import ElementsWalletProvider from '../../lib/wallet/providers/ElementsWalletProvider';
import { bitcoinClient, elementsClient } from './Nodes';

describe('Core', () => {
  const bip32 = BIP32Factory(ecc);
  const slip77 = SLIP77Factory(ecc);

  const database = new Database(Logger.disabledLogger, ':memory:');

  let wallet: Wallet;
  let walletLiquid: WalletLiquid;

  beforeAll(async () => {
    await Promise.all([setup(), database.init()]);

    const initWallet = (w: Wallet, network: any) => {
      w.initKeyProvider(
        network,
        'm/0/0',
        0,
        bip32.fromSeed(mnemonicToSeedSync(generateMnemonic())),
      );
    };

    wallet = new Wallet(
      Logger.disabledLogger,
      CurrencyType.BitcoinLike,
      new CoreWalletProvider(Logger.disabledLogger, bitcoinClient),
    );
    initWallet(wallet, Networks.bitcoinRegtest);

    walletLiquid = new WalletLiquid(
      Logger.disabledLogger,
      new ElementsWalletProvider(Logger.disabledLogger, elementsClient),
      slip77.fromSeed(generateMnemonic()),
    );
    initWallet(walletLiquid, networks.regtest);

    await bitcoinClient.connect();
    await elementsClient.connect();

    // To have a blinded output in the Elements wallet
    await elementsClient.sendToAddress(
      await elementsClient.getNewAddress(),
      10 ** 8,
    );

    await Promise.all([bitcoinClient.generate(1), elementsClient.generate(1)]);
  });

  afterAll(async () => {
    await Promise.all([bitcoinClient.generate(1), elementsClient.generate(1)]);

    bitcoinClient.disconnect();
    elementsClient.disconnect();
  });

  test('should get output value of Bitcoin transactions', async () => {
    const outputAmount = 420234;
    const { transaction, vout } = await wallet.sendToAddress(
      await wallet.getAddress(),
      outputAmount,
    );

    expect(getOutputValue(wallet, transaction!.outs[vout!])).toEqual(
      outputAmount,
    );
  });

  test('should get output value of unblinded Liquid transactions', async () => {
    const outputAmount = 562312;
    const { transaction, vout } = await walletLiquid.sendToAddress(
      fromOutputScript(
        CurrencyType.Liquid,
        toOutputScript(
          CurrencyType.Liquid,
          await walletLiquid.getAddress(),
          networks.regtest,
        ),
        networks.regtest,
      ),
      outputAmount,
    );

    expect(
      getOutputValue(
        walletLiquid,
        (transaction as LiquidTransaction).outs[vout!],
      ),
    ).toEqual(outputAmount);

    // Wrong asset
    walletLiquid['network'] = networks.liquid;
    expect(
      getOutputValue(
        walletLiquid,
        (transaction as LiquidTransaction).outs[vout!],
      ),
    ).toEqual(0);
    walletLiquid['network'] = networks.regtest;
  });

  test('should get output value of blinded Liquid transactions', async () => {
    const script = toOutputScript(
      CurrencyType.Liquid,
      await walletLiquid.getAddress(),
      walletLiquid.network,
    );

    const outputAmount = 1245412;
    const { transaction, vout } = await walletLiquid.sendToAddress(
      walletLiquid.encodeAddress(script),
      outputAmount,
    );

    expect(
      getOutputValue(
        walletLiquid,
        (transaction as LiquidTransaction).outs[vout!],
      ),
    ).toEqual(outputAmount);

    // Wrong asset hash
    walletLiquid['network'] = networks.liquid;
    expect(
      getOutputValue(
        walletLiquid,
        (transaction as LiquidTransaction).outs[vout!],
      ),
    ).toEqual(0);
    walletLiquid['network'] = networks.regtest;
  });

  test('should construct legacy claim details', async () => {
    const preimage = randomBytes(32);
    const claimKeys = wallet.getNewKeys();
    const refundKeys = ECPair.makeRandom();

    const redeemScript = swapScript(
      crypto.sha256(preimage),
      claimKeys.keys.publicKey,
      refundKeys.publicKey,
      2,
    );
    const outputScript = p2wshOutput(redeemScript);

    const tx = Transaction.fromHex(
      await bitcoinClient.getRawTransaction(
        await bitcoinClient.sendToAddress(
          wallet.encodeAddress(outputScript),
          100_00,
        ),
      ),
    );
    const output = detectSwap(redeemScript, tx);

    const claimDetails = constructClaimDetails(
      {
        get: jest.fn().mockReturnValue(OutputType.Bech32),
      } as unknown as SwapOutputType,
      wallet,
      {
        keyIndex: claimKeys.index,
        version: SwapVersion.Legacy,
        redeemScript: getHexString(redeemScript),
      } as unknown as Swap,
      tx,
      preimage,
    ) as ClaimDetails;

    expect(claimDetails).toEqual({
      ...output,
      preimage,
      redeemScript,
      txHash: tx.getHash(),
      type: OutputType.Bech32,
      keys: wallet.getKeysByIndex(claimKeys.index),
    });
  });

  test('should construct Taproot claim details', async () => {
    const preimage = randomBytes(32);
    const claimKeys = wallet.getNewKeys();
    const refundKeys = ECPair.makeRandom();

    const tree = swapTree(
      false,
      crypto.sha256(preimage),
      claimKeys.keys.publicKey,
      refundKeys.publicKey,
      2,
    );
    const musig = createMusig(claimKeys.keys, refundKeys.publicKey);
    const tweakedKey = tweakMusig(CurrencyType.BitcoinLike, musig, tree);
    const outputScript = p2trOutput(tweakedKey);

    const tx = Transaction.fromHex(
      await bitcoinClient.getRawTransaction(
        await bitcoinClient.sendToAddress(
          wallet.encodeAddress(outputScript),
          100_00,
        ),
      ),
    );
    const output = detectSwap(tweakedKey, tx)!;

    const claimDetails = constructClaimDetails(
      {} as unknown as SwapOutputType,
      wallet,
      {
        keyIndex: claimKeys.index,
        version: SwapVersion.Taproot,
        lockupTransactionVout: output.vout,
        refundPublicKey: refundKeys.publicKey,
        redeemScript: SwapTreeSerializer.serializeSwapTree(tree),
      } as unknown as Swap,
      tx,
      preimage,
    ) as ClaimDetails;

    expect(claimDetails.vout).toEqual(output.vout);
    expect(claimDetails.preimage).toEqual(preimage);
    expect(claimDetails.value).toEqual(output.value);
    expect(claimDetails.keys).toEqual(claimKeys.keys);
    expect(claimDetails.txHash).toEqual(tx.getHash());
    expect(claimDetails.script).toEqual(output.script);
    expect(claimDetails.type).toEqual(OutputType.Taproot);
    expect(claimDetails.cooperative).toEqual(false);
    expect(claimDetails.internalKey).toEqual(musig.getAggregatedPublicKey());
    expect(
      SwapTreeSerializer.serializeSwapTree(claimDetails.swapTree!),
    ).toEqual(SwapTreeSerializer.serializeSwapTree(tree));
  });

  test('should create Musig', () => {
    const ourKeys = ECPair.fromPrivateKey(
      getHexBuffer(
        'a01c77e7effc4d49dd2d2f345b9dde1984c426b17c6756d10041819f73856b8e',
      ),
    );
    const theirPublicKey = getHexBuffer(
      '03d8b9ee7742d8f98484f8e3826083d13453547dea7e8f2976f5733672f3a71f4a',
    );
    const musig = createMusig(ourKeys, theirPublicKey);

    expect(musig.publicKeys).toEqual([ourKeys.publicKey, theirPublicKey]);
    expect(musig.getAggregatedPublicKey()).toMatchSnapshot();
  });

  test.each`
    isLiquid
    ${true}
    ${false}
  `('should tweak Musig (liquid: $isLiquid)', ({ isLiquid }) => {
    const tree = SwapTreeSerializer.deserializeSwapTree({
      claimLeaf: {
        version: 192,
        output:
          '82012088a91412fb93c968f24c233c64081fb734e34e9ecb0cfb88206584eb8c9a539e24fed7a81b002891506ee94fb32ee7c6c23a488caeedd5a653ac',
      },
      refundLeaf: {
        version: 192,
        output:
          '20fb00c6a397ac4f25c6aa85392ded2e3487f219a150ef7ac615f33310b1d95945ad025e07b1',
      },
    });

    const ourKeys = ECPair.fromPrivateKey(
      getHexBuffer(
        'a01c77e7effc4d49dd2d2f345b9dde1984c426b17c6756d10041819f73856b8e',
      ),
    );
    const theirPublicKey = getHexBuffer(
      '03d8b9ee7742d8f98484f8e3826083d13453547dea7e8f2976f5733672f3a71f4a',
    );
    const musig = createMusig(ourKeys, theirPublicKey);
    const tweakedMusig = tweakMusig(
      isLiquid ? CurrencyType.Liquid : CurrencyType.BitcoinLike,
      musig,
      tree,
    );

    expect(tweakedMusig).toMatchSnapshot();
  });

  test('should hash Bitcoin transactions for witness v1', async () => {
    const keys = ECPair.makeRandom();

    const outputScript = Scripts.p2trOutput(toXOnly(keys.publicKey));
    const amountSent = 100_000;

    const tx = parseTransaction(
      CurrencyType.BitcoinLike,
      await bitcoinClient.getRawTransaction(
        await bitcoinClient.sendToAddress(
          address.fromOutputScript(outputScript, Networks.bitcoinRegtest),
          amountSent,
        ),
      ),
    );
    const outputToSpendIndex = tx.outs.findIndex((output) =>
      output.script.equals(outputScript),
    );

    const spendTx = new Transaction();
    spendTx.addInput(tx.getHash(), outputToSpendIndex);
    spendTx.addOutput(
      address.toOutputScript(
        await bitcoinClient.getNewAddress(),
        Networks.bitcoinRegtest,
      ),
      amountSent - 1_000,
    );

    const hash = await hashForWitnessV1(
      {
        chainClient: bitcoinClient,
        type: CurrencyType.BitcoinLike,
        network: Networks.bitcoinRegtest,
      } as Currency,
      spendTx,
      0,
    );

    spendTx.setWitness(0, [keys.signSchnorr(hash)]);

    await bitcoinClient.sendRawTransaction(spendTx.toHex());
  });

  test('should hash Liquid transactions for witness v1', async () => {
    const keys = ECPair.makeRandom();

    const outputScript = Scripts.p2trOutput(toXOnly(keys.publicKey));
    const amountSent = 100_000;

    const tx = parseTransaction(
      CurrencyType.Liquid,
      await elementsClient.getRawTransaction(
        await elementsClient.sendToAddress(
          address.fromOutputScript(outputScript, networks.regtest),
          amountSent,
        ),
      ),
    ) as LiquidTransaction;
    const outputToSpendIndex = tx.outs.findIndex((output) =>
      output.script.equals(outputScript),
    );

    const pset = Creator.newPset();
    pset.addInput(
      new CreatorInput(tx.getId(), outputToSpendIndex).toPartialInput(),
    );
    pset.inputs[0].sighashType = LiquidTransaction.SIGHASH_DEFAULT;

    const feeAmount = 1_000;

    const updater = new Updater(pset);
    updater.addInWitnessUtxo(0, tx.outs[outputToSpendIndex]);
    updater.addOutputs([
      {
        amount: amountSent - feeAmount,
        asset: LiquidNetworks.liquidRegtest.assetHash,
        script: liquidAddress.toOutputScript(
          await elementsClient.getNewAddress(),
        ),
      },
      {
        amount: feeAmount,
        asset: LiquidNetworks.liquidRegtest.assetHash,
      },
    ]);

    const finalizer = new Finalizer(pset);
    finalizer.finalizeInput(0, () => {
      return {
        finalScriptWitness: Buffer.alloc(64),
      };
    });

    finalizer.finalize();
    const spendTx = Extractor.extract(pset);

    const hash = await hashForWitnessV1(
      {
        chainClient: elementsClient,
        type: CurrencyType.Liquid,
        network: LiquidNetworks.liquidRegtest,
      } as unknown as Currency,
      spendTx,
      0,
    );

    spendTx.setWitness(0, [keys.signSchnorr(hash)]);

    await elementsClient.sendRawTransaction(spendTx.toHex());
  });
});
