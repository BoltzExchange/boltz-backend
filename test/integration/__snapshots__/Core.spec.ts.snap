// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core should create Musig 1`] = `
{
  "data": [
    87,
    154,
    31,
    116,
    125,
    206,
    223,
    181,
    56,
    159,
    20,
    192,
    226,
    22,
    37,
    227,
    97,
    252,
    190,
    142,
    233,
    192,
    140,
    221,
    128,
    221,
    96,
    76,
    118,
    96,
    179,
    195,
  ],
  "type": "Buffer",
}
`;

exports[`Core should tweak Musig (liquid: false) 1`] = `
{
  "data": [
    134,
    128,
    242,
    101,
    31,
    54,
    59,
    145,
    216,
    96,
    179,
    41,
    22,
    75,
    0,
    115,
    85,
    178,
    193,
    200,
    250,
    194,
    117,
    88,
    194,
    37,
    92,
    49,
    105,
    243,
    91,
    4,
  ],
  "type": "Buffer",
}
`;

exports[`Core should tweak Musig (liquid: true) 1`] = `
{
  "data": [
    9,
    247,
    149,
    248,
    81,
    130,
    205,
    22,
    53,
    2,
    241,
    64,
    215,
    138,
    228,
    40,
    231,
    22,
    79,
    142,
    169,
    21,
    148,
    3,
    200,
    221,
    32,
    95,
    98,
    129,
    20,
    141,
  ],
  "type": "Buffer",
}
`;
